File: AlgoExpert/two-number-sum.go
package main

import "sort"

// O(n^2)
func TwoNumberSum_n2(array []int, target int) []int {
	arrayLength := len(array)
	for i, v := range array {
		complement := target - v
		for j := i + 1; j < arrayLength; j++ {
			if array[j] == complement {
				return []int{v, array[j]}
			}
		}
	}
	return make([]int, 0)
}

// O(n)
func TwoNumberSum_n(array []int, target int) []int {
	hashmap := make(map[int]int)
	for i, v := range array {
		complement := target - v
		if _, ok := hashmap[complement]; ok {
			return []int{v, complement}
		}
		hashmap[v] = i
	}
	return []int{}
}

// O(nlog(n))
func TwoNumberSum_nlogn(array []int, target int) []int {
	arrayLength := len(array)
	left, right := 0, arrayLength-1
	sort.Ints(array)
	for left < right {
		leftElement, rightElement := array[left], array[right]
		total := leftElement + rightElement
		if total == target {
			return []int{leftElement, rightElement}
		} else if total > target {
			right--
		} else {
			left++
		}
	}
	return []int{}
}

func TwoNumberSum(array []int, target int) []int {
	return TwoNumberSum_n(array, target)
}

File: AlgoExpert/two-number-sum.py
# O(n^2)
def twoNumberSum_n2(array, targetSum):
    for x in array:
        complement = targetSum - x
        for y in array:
            if x == y:
                continue
            if y == complement:
                return [x, y]
    return []


# O(n)
def twoNumberSum_n(array, targetSum):
    complement_store = {}
    for x in array:
        complement = targetSum - x
        if complement_store.get(complement) is not None:
            return [x, complement]
        complement_store[x] = complement
    return []


# O(nlog(n))
def twoNumberSum_nlogn(array, targetSum):
    sorted_array = list(sorted(array))
    left, right = 0, len(array) - 1
    while left < right:
        n1, n2 = sorted_array[left], sorted_array[right]
        total = n1 + n2
        if total == targetSum:
            return [n1, n2]
        elif total > targetSum:
            right -= 1
        else:
            left += 1
    return []


def twoNumberSum_n2(array, targetSum):
    return twoNumberSum_n(array, targetSum)
